### ISSUE TITLE
tictactoeBrain tries to lock pthread mutex before initialization 

### ISSUE THREAD

**Issue by shrine79** [ScazLab/baxter_tictactoe#22]
@alecive I tried the baxter tictactoe package on Baxter and encountered an issue during calibration stage. The tictactoeBrain node randomly gets stuck in its the constructor and the process doesn't exit gracefully after Ctrl-c is pressed. 

I think there is a problem that the "setBrainState()" function tries to lock "_mutex_brain", which is a pthread mutex, in the class constructor before it's properly initialized. 

Quote from http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html

"If mutex does not refer to an initialized mutex object, the behavior of pthread_mutex_lock(), pthread_mutex_trylock(), and pthread_mutex_unlock() is undefined."

**Comment by alecive** [comment]
Interesting. This never happened to me; neither with or without the robot (I just tested in both cases). Overall, this could be the right occasion to move out from POSIX threads and use more andvanced C++11 features such as shared pointers.

I just moved initialization of the mutexes `after` setBrain state (see https://github.com/ScazLab/baxter_tictactoe/commit/04f24309165131de6f6daccb89ed703f6bb0ad19 ), can you please test it on your robot and report to me if that does not work?

Also, feel free to contribute to a README of sorts on how to run the demo. We are the only users right now and it's difficult to gauge the README's efficacy without external users.

**Comment by shrine79** [comment]
@alecive I've tested the solution on Ubuntu 14.04 and ROS Indigo before I created the issue. The problem has never occurred in the first run in a brand new terminal, but would occur in a indeterminate manner if the group of nodes are killed (Ctrl-C) and relaunch in the same terminal. I could sometime find "roslaunch" process is still running after Ctrl-C with ps command. The sensor calibration usually takes 30 seconds to forever before the solution. It's literally done immediately after the fix. 

My guess is that the pthread runtime is trying to recycle used mutex objects. Locking an uninitialized mutex may succeed but would possibly create issues on its destruction.  This in turn would fail the lock attempt next time and stop the node in its constructor. 

Quote: "The pthread_mutex_destroy() function shall destroy the mutex object referenced by mutex; the mutex object becomes, in effect, uninitialized. An implementation may cause pthread_mutex_destroy() to set the object referenced by mutex to an invalid value. A destroyed mutex object can be reinitialized using pthread_mutex_init(); the results of otherwise referencing the object after it has been destroyed are undefined. "

I am a student from University of Canberra and I've finished my study on Baxter. New students are coming to take over the project and I'll make them involved in improving README.

**Comment by alecive** [comment]
@shrine79 sorry if I don't understand, so now it works? and we can close the issue?

Another question, what do you mean with "sensor calibration"?


> I am a student from University of Canberra and I've finished my study on Baxter. New students are coming to take over the project and I'll make them involved in improving README.

That would be great!

**Comment by alecive** [comment]
Also, if you want to request new features, feel free to ask them, we can easily integrate everything into the main repository (e.g. make the user move first instead of the robot, as you did in your personal copy of this repo).

**Comment by shrine79** [comment]
@alecive Sorry to make you confused. I had the same fix as you did in my workspace and tested before I created the issue. It works with the fix and issue can be closed. 

I used the term "sensor calibration" to refer the STATE_CALIB state in board_state_sensor node. The board_state_sensor node would not finish calibration until tictactoeBrain node reaches certain state. So the Baxter looks like doing sensor calibration forever.
